### PR TITLE
fix(animations): generate aot code for animation trigger output events

### PR DESCRIPTION
**Pull request by matsko**

### CODE REVIEW

**Comment by matsko** [comment]
@tbosch `non jit` code fails for animation integration tests.

Also, we now explose `player` within animations (which I think is a useful feature to have since it allows for users to get access to the player to pause and cancel the animation programmatically).

**Comment by matsko** [comment]
Aot Sample:

```
    _View_App0.prototype.detectChangesInternal = function(throwOnChange) {
        var self = this;
        var transitionAnimation = null ;
        var currVal_4 = self.context.openClose;
        if (jit_checkBinding6(throwOnChange, self._expr_4, currVal_4)) {
            var oldRenderVar = self._expr_4;
            if ((oldRenderVar == jit_UNINITIALIZED5) ) {
                (oldRenderVar = 'void');
            }
            var newRenderVar = currVal_4;
            if ((newRenderVar == jit_UNINITIALIZED5) ) {
                (newRenderVar = 'void');
            }
            transitionAnimation = self.componentType.animations['myTrigger'](self, self._el_8, oldRenderVar, newRenderVar);
            self._expr_4 = currVal_4;
        }
        if (transitionAnimation) {
            transitionAnimation.player.onDone(self.eventHandler(self._handle_myTrigger_8_0.bind(self, transitionAnimation)));
        }
        if (transitionAnimation) {
            transitionAnimation.player.onStart(self.eventHandler(self._handle_myTrigger_8_1.bind(self, transitionAnimation)));
        }
        self.detectContentChildrenChanges(throwOnChange);
        self.detectViewChildrenChanges(throwOnChange);
    }
```

**Comment by matsko** [comment]
main refactorings:
- generate events within onDone and onStart
- return only player from animation factory
- register inside of the if block

**Comment by matsko** [comment]
Codegen looks like this now:

```js
    ;
    _View_App0.prototype.detectChangesInternal = function(throwOnChange) {
        var self = this;
        var currVal_4 = self.context.openClose;
        if (jit_checkBinding6(throwOnChange, self._expr_4, currVal_4)) {
            var oldRenderVar = self._expr_4;
            if ((oldRenderVar == jit_UNINITIALIZED5) ) {
                (oldRenderVar = 'void');
            }
            var newRenderVar = currVal_4;
            if ((newRenderVar == jit_UNINITIALIZED5) ) {
                (newRenderVar = 'void');
            }
            var animationTransition_myTrigger = self.componentType.animations['myTrigger'](self, self._el_8, oldRenderVar, newRenderVar);
            animationTransition_myTrigger.onDone(self._handle_myTrigger_8_0.bind(self));
            animationTransition_myTrigger.onStart(self._handle_myTrigger_8_1.bind(self));
            self._expr_4 = currVal_4;
        }
        self.detectContentChildrenChanges(throwOnChange);
        self.detectViewChildrenChanges(throwOnChange);
    }
    ;
```

File: modules/@angular/compiler/src/view_compiler/property_binder.ts [hunk 1]
```diff
@@ -172,12 +176,32 @@ function bindAndWriteToRenderer(
             [newRenderVar.set(emptyStateValue).toStmt()]));
```
Comment (tbosch) [inline modules/@angular/compiler/src/view_compiler/property_binder.ts]: Remove the variables `oldRenderVar` and `newRenderVar`, and use a ternary expression instead.
```
...animationFnExpr.callFn([o.THIS_EXPR,
  oldRenderValue.equals(o.importExpr(resolveIdentifier(Identifiers.UNINITIALIZED))
    .conditional(oldRenderVar, emptyStateValue)),
  newRenderValue.equals(o.importExpr(resolveIdentifier(Identifiers.UNINITIALIZED))
    .conditional(newRenderVar, emptyStateValue)),
...
```

File: modules/@angular/compiler/src/animation/animation_compiler.ts [hunk 2]
```diff
@@ -290,7 +292,7 @@ class _AnimationBuilder implements AnimationAstVisitor {
           new o.FnParam(_ANIMATION_CURRENT_STATE_VAR.name, o.DYNAMIC_TYPE),
           new o.FnParam(_ANIMATION_NEXT_STATE_VAR.name, o.DYNAMIC_TYPE)
         ],
-        statements);
+        statements, o.DYNAMIC_TYPE);
```
Comment (tbosch) [inline modules/@angular/compiler/src/animation/animation_compiler.ts]: This should be `o.importType(resolveIdentifier(Identifiers.AnimationTransition))`, as the factory returns an `AnimationTransition`, right?

File: modules/@angular/compiler/src/view_compiler/event_binder.ts [hunk 3]
```diff
@@ -10,6 +10,7 @@ import {CompileDirectiveMetadata} from '../compile_metadata';
 import {isPresent} from '../facade/lang';
 import {identifierToken} from '../identifiers';
 import * as o from '../output/output_ast';
+import {ViewType} from '../private_import_core';
```
Comment (tbosch) [inline modules/@angular/compiler/src/view_compiler/event_binder.ts]: Not needed

File: modules/@angular/compiler/src/view_compiler/property_binder.ts [hunk 4]
```diff
@@ -148,8 +149,11 @@ function bindAndWriteToRenderer(
           targetViewExpr = compileElement.appElement.prop('componentView');
         }
 
+        var detachStmts: o.Statement[] = [];
         compileMethod = view.animationBindingsMethod;
 
+        const _ANIMATION_TRANSITION_VAR = o.variable('animationTransition_' + animationName);
```
Comment (tbosch) [inline modules/@angular/compiler/src/view_compiler/property_binder.ts]: remove `_` prefix, and don't upper case everything. This is just a variable here, not a constant defined at the top leve.

Comment (tbosch) [inline modules/@angular/compiler/src/view_compiler/property_binder.ts]: Move this further down, right before you declare the variable (as it is not used before that).

File: modules/@angular/compiler/src/view_compiler/property_binder.ts [hunk 5]
```diff
@@ -172,12 +176,32 @@ function bindAndWriteToRenderer(
...
         updateStmts.push(
-            animationFnExpr.callFn([o.THIS_EXPR, renderNode, oldRenderVar, newRenderVar]).toStmt());
+            _ANIMATION_TRANSITION_VAR
+                .set(animationFnExpr.callFn([o.THIS_EXPR, renderNode, oldRenderVar, newRenderVar]))
+                .toDeclStmt());
 
-        view.detachMethod.addStmt(
-            animationFnExpr.callFn([o.THIS_EXPR, renderNode, oldRenderValue, emptyStateValue])
-                .toStmt());
+        detachStmts.push(_ANIMATION_TRANSITION_VAR
```
Comment (tbosch) [inline modules/@angular/compiler/src/view_compiler/property_binder.ts]: Why are we always creating the animation again on view detach?

File: modules/@angular/compiler/src/view_compiler/property_binder.ts [hunk 6]
```diff
@@ -172,12 +176,32 @@ function bindAndWriteToRenderer(
...
+
+        eventListeners.forEach(listener => {
+          if (listener.isAnimation && listener.eventName === animationName) {
+            let callbackMethod = listener.eventPhase == 'start' ? 'onStart' : 'onDone';
+            let transitionEvent = o.variable('e');
+            let outputStmt =
+                _ANIMATION_TRANSITION_VAR
+                    .callMethod(
+                        callbackMethod,
+                        [o.THIS_EXPR.prop(listener.methodName).callMethod('bind', [o.THIS_EXPR])])
```
Comment (tbosch) [inline modules/@angular/compiler/src/view_compiler/property_binder.ts]: Use `.callMethod(o.BuiltinMethod.Bind, ...`, don't pass in `bind` as a string.

File: modules/@angular/compiler/src/view_compiler/property_binder.ts [hunk 7]
```diff
@@ -172,12 +176,32 @@ function bindAndWriteToRenderer(
...
-            animationFnExpr.callFn([o.THIS_EXPR, renderNode, oldRenderValue, emptyStateValue])
-                .toStmt());
+        detachStmts.push(_ANIMATION_TRANSITION_VAR
+                             .set(animationFnExpr.callFn(
+                                 [o.THIS_EXPR, renderNode, oldRenderValue, emptyStateValue]))
+                             .toDeclStmt());
+
+        eventListeners.forEach(listener => {
+          if (listener.isAnimation && listener.eventName === animationName) {
+            let callbackMethod = listener.eventPhase == 'start' ? 'onStart' : 'onDone';
```
Comment (tbosch) [inline modules/@angular/compiler/src/view_compiler/property_binder.ts]: Move the logic here into `CompileEventListener.listenToAnimation`.

File: modules/@angular/core/src/animation/animation_transition.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+import {AnimationPlayer} from './animation_player';
+import {AnimationTransitionEvent} from './animation_transition_event';
+
+export class AnimationTransition {
+  private _fromState: string;
+  private _toState: string;
+  private _totalTime: number;
+
+  constructor(
+      public player: AnimationPlayer,
```
Comment (tbosch) [inline modules/@angular/core/src/animation/animation_transition.ts]: Why does the player need to be public?

File: modules/@angular/core/src/animation/animation_transition.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+      {fromState, toState, totalTime}: {fromState: string, toState: string, totalTime: number}) {
```
Comment (tbosch) [inline modules/@angular/core/src/animation/animation_transition.ts]: Don't use named arguments here, just positional ones and declare the fields as well. E.g.
```
constructor(private _player: AnimationPlayer, private _fromState: string, ...) {}
```

File: tools/public_api_guard/core/index.d.ts [hunk 10]
```diff
@@ -109,12 +109,14 @@ export declare class AnimationStyleMetadata extends AnimationMetadata {
...
     fromState: string;
+    phaseName: string;
     toState: string;
     totalTime: number;
-    constructor({fromState, toState, totalTime}: {
+    constructor({fromState, toState, totalTime, phaseName}: {
         fromState: string;
         toState: string;
         totalTime: number;
+        phaseName: string;
```
Comment (tbosch) [inline tools/public_api_guard/core/index.d.ts]: Please ping @IgorMinar whether this is a breaking change, and we need to make this field optional. I don't expect users to new up their own `AnimationTransitionEvent`.

File: modules/@angular/core/src/animation/animation_transition.ts [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+import {AnimationPlayer} from './animation_player';
+import {AnimationTransitionEvent} from './animation_transition_event';
+
+export class AnimationTransition {
+  constructor(
+      private _player: AnimationPlayer, private _fromState: string, private _toState: string,
+      private _totalTime: number) {}
+
+  /** @internal */
+  _createEvent(phaseName: string): AnimationTransitionEvent {
```
Comment (tbosch) [inline modules/@angular/core/src/animation/animation_transition.ts]: why not `private`

**Pull request merged**